ql/time/daycounters/actualactual.cpp

```diff
@@ -40,111 +44,120 @@ namespace QuantLib {
-            } else {
-                // here refPeriodStart is the next (maybe notional)
-                // payment date and refPeriodEnd is the second next
-                // (maybe notional) payment date.
-                // d1 < refPeriodStart < refPeriodEnd
-                // AND d2 <= refPeriodEnd
-                // this case is long first coupon
-
-                // the last notional payment date
-                Date previousRef;
-                if (schedule_.empty()) {
-                    previousRef = refPeriodStart - months*Months;
-                } else {
-                    previousRef = schedule_.calendar().advance(refPeriodStart,
-                                                               -schedule_.tenor(),
-                                                               schedule_.businessDayConvention(),
-                                                               schedule_.endOfMonth());
-                }
-
-                if (d2 > refPeriodStart)
-                    return yearFraction(d1, refPeriodStart, previousRef,
-                                        refPeriodStart) +
-                        yearFraction(refPeriodStart, d2, refPeriodStart,
-                                     refPeriodEnd);
-                else
-                    return yearFraction(d1,d2,previousRef,refPeriodStart);
-            }
-        } else {
-            // here refPeriodEnd is the last (notional?) payment date
-            // d1 < refPeriodEnd < d2 AND refPeriodStart < refPeriodEnd
-            QL_REQUIRE(refPeriodStart<=d1,
-                       "invalid dates: "
-                       "d1 < refPeriodStart < refPeriodEnd < d2");
-            // now it is: refPeriodStart <= d1 < refPeriodEnd < d2
-
-            // the part from d1 to refPeriodEnd
-            Time sum = yearFraction(d1, refPeriodEnd,
-                                    refPeriodStart, refPeriodEnd);
-
-            // the part from refPeriodEnd to d2
-            // count how many regular periods are in [refPeriodEnd, d2],
-            // then add the remaining time
-            Integer i=0;
-            Date newRefStart, newRefEnd;
-            for (;;) {
-                newRefStart = refPeriodEnd + (months*i)*Months;
-                newRefEnd = refPeriodEnd + (months*(i+1))*Months;
-                if (d2 < newRefEnd) {
-                    break;
-                } else {
-                    sum += period;
-                    i++;
-                }
-            }
-            sum += yearFraction(newRefStart,d2,newRefStart,newRefEnd);
-            return sum;
-        }
+		// Base Cases;
+		if (d1 == d2) {
+			return 0.0;
+		} else if (d2 < d1) {
+			return -yearFraction(d2, d1, d3, d4);
+		}
+
+		//To account for the fact that some higher level classes to not work out the reference periods correctly.
+		//A day counter with a schedule takes precedence over an explicit reference period.
```

Comment:
My reasoning was that I feel like having all four dates is a bit of a crutch. As far as I could tell it is only relevant for the actual actual day counters, and means that you have to add a lot of logic to the bond class to try to work out the correct reference periods, when that really belongs in the ActualActual class. So this seemed like a logical first step towards deprecating the four date usage. 

I think this fix has the potential to simplify the day counters by making all calls to them only every require the start and end date.

Reply:
(I appreciate that I dont necessarily understand every use case for the ActualActual, but it seems like that day count convention just always needs a schedule as a matter of definition, so ideally i would like to see the constructor require a schedule, and if you do that then it can always work out its own reference dates, so I don't see why at that point you would every use a four date version of the function).